File: ios/Jazzify/Defense/Engine/DefenseStaffChordLabel.swift
import Foundation

enum DefenseStaffChordLabel {
    static func labels(for chord: SurvivalPhraseChord) -> [String] {
        let steps = SurvivalPhraseChordSteps.getSteps(notes: chord.notes)
        let hasStaffChordNames = chord.notes.contains {
            !($0.staffChordName?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
        }
        var previousStaffLabel = ""
        return steps.enumerated().map { stepPosition, step in
            var stepStaffLabel = ""
            for noteIndex in step.noteIndices {
                guard noteIndex < chord.notes.count else { continue }
                if let label = chord.notes[noteIndex].staffChordName?.trimmingCharacters(in: .whitespacesAndNewlines),
                   !label.isEmpty {
                    stepStaffLabel = label
                }
            }
            if hasStaffChordNames {
                if !stepStaffLabel.isEmpty, stepStaffLabel != previousStaffLabel {
                    previousStaffLabel = stepStaffLabel
                    return stepStaffLabel
                }
                return ""
            }
            if stepPosition == 0 {
                return chord.chordName
            }
            return ""
        }
    }

    static func label(for chord: SurvivalPhraseChord, stepIndex: Int) -> String {
        let all = labels(for: chord)
        guard stepIndex >= 0, stepIndex < all.count else { return "" }
        return all[stepIndex]
    }
}

File: src/game/defense/defenseStaffChordLabel.ts
/**
 * Staff chord label resolution for defense phrase chords (matches buildDefenseStaffGroups).
 */
import type { DefensePhraseChord } from '@/game/defense/defenseTypes';
import { getPhraseChordSteps } from '@/utils/phraseChordSteps';

export const resolveDefenseStaffChordLabels = (
  chord: DefensePhraseChord,
): readonly string[] => {
  const { steps } = getPhraseChordSteps(chord.notes);
  const hasStaffChordNames = chord.notes.some(
    (note) => (note.staffChordName ?? '').trim().length > 0,
  );
  let previousStaffLabel = '';
  return steps.map((step, stepPosition) => {
    let stepStaffLabel = '';
    for (const noteIndex of step.noteIndices) {
      const note = chord.notes[noteIndex];
      const noteStaffLabel = (note?.staffChordName ?? '').trim();
      if (noteStaffLabel.length > 0) {
        stepStaffLabel = noteStaffLabel;
      }
    }

    if (hasStaffChordNames) {
      if (stepStaffLabel.length > 0 && stepStaffLabel !== previousStaffLabel) {
        previousStaffLabel = stepStaffLabel;
        return stepStaffLabel;
      }
      return '';
    }
    if (stepPosition === 0) {
      return chord.chordName;
    }
    return '';
  });
};

export const resolveDefenseStaffChordLabelForStep = (
  chord: DefensePhraseChord,
  stepIndex: number,
): string => resolveDefenseStaffChordLabels(chord)[stepIndex] ?? '';
